### PR TITLE
Skip full-vocabulary logits the backbone never reads (logits_to_keep=1)

### PR DESCRIPTION
**Pull request by dvdimitrov13**
## What

`Qwen3Backbone.forward` now passes `logits_to_keep=1`, confining the `lm_head`
matmul to a single position. One argument; the rest of the diff is a test.

## Why

The backbone consumes only `hidden_states[-1]`, but
`Qwen3VLForConditionalGeneration.forward` evaluates `lm_head` unconditionally:

```python
slice_indices = slice(-logits_to_keep, None) if isinstance(logits_to_keep, int) else logits_to_keep
logits = self.lm_head(hidden_states[:, slice_indices, :])
```

With the default `logits_to_keep=0`, `slice(-0, None)` is `slice(0, None)` — the
whole sequence. For Cosmos-Reason2-2B that is a `2048 → 151936` GEMM over every
token, producing a `[B, S, 151936]` tensor that is discarded on the next line.
The `logits_to_keep` parameter exists for exactly this case ("Only compute
necessary logits") and was simply never passed.

## Measured (RTX 5090, torch 2.9.0+cu128, bf16)

Isolated `lm_head` at production dimensions, full sequence vs `logits_to_keep=1`:

| scenario | full | `logits_to_keep=1` | saved |
|---|---|---|---|
| inference B=1, S=160 | 0.673 ms | 0.380 ms | **0.29 ms** |
| inference B=1, S=320 | 1.064 ms | 0.380 ms | **0.68 ms** |
| finetune B=8, S=160 | 3.351 ms | 0.379 ms | **2.97 ms** |
| finetune B=32, S=160 | 13.416 ms | 0.462 ms | **12.95 ms** |
| finetune B=32, S=320 | 26.842 ms | 0.523 ms | **26.32 ms** |

Discarded activation removed, `B × S × 151936 × 2` bytes: **1.56 GB** at
B=32/S=160, **3.11 GB** at B=32/S=320 (double that if logits are upcast to fp32).

This is mostly a **fine-tuning** win — memory and step time at the documented
default `GLOBAL_BATCH_SIZE=32`. Inference gains ~0.3 ms, well under 1% of the
backbone, because the waste scales with `batch × sequence` and batch-1 inference
has little of it. Reporting that plainly rather than as a headline.

Note the `logits_to_keep=1` column is flat at ~0.38 ms regardless of sequence
length: that is the 622 MB weight matrix being streamed (622 MB / 0.38 ms ≈
1.64 TB/s ≈ the card's bandwidth). Avoiding that too would mean not calling
`lm_head` at all, which is a structural change; happy to follow up separately if
you want the extra ~0.38 ms.

## Losslessness

Every entry of `hidden_states` is **bitwise unchanged**; only the unread `logits`
shrink, from `[B, S, vocab]` to `[B, 1, vocab]`. Verified on a real (tiny)
Qwen3-VL, not argued from the source.

## Why not just call the inner `Qwen3VLModel`?

Skipping `lm_head` entirely by reading `self.model.model(...).last_hidden_state`
looks equivalent and **is not**:

- `Qwen3VLModelOutputWithPast` **has** `last_hidden_state`, so
  `check_model_inputs(tie_last_hidden_states=True)` ties `hidden_states[-1]` to the
  **post**-final-norm tensor;
- `Qwen3VLCausalLMOutputWithPast` has no such field, so the wrapper's
  `hidden_states[-1]` is the **pre**-final-norm decoder output this checkpoint was
  trained against.

Measured on a tiny Qwen3-VL: wrapper vs inner `max_abs_diff = 3.25`, while
`inner.last_hidden_state == language_model.norm(wrapper.hidden_states[-1])` exactly
(diff `0.0`). Switching would silently insert an RMSNorm into the feature path, so
`test_pre_norm_semantics_are_load_bearing` pins this.

## Tests

`tests/gr00t/model/test_backbone_lm_head_skipped.py`, 5 tests, CPU-only on a tiny
randomly-initialised Qwen3-VL — no gated checkpoint download, no GPU:

- `lm_head` sees one position, not the sequence
- every `hidden_states` entry bitwise identical (the safety claim)
- only the unread logits shrink
- `Qwen3Backbone.forward` actually forwards the argument
- the pre-norm semantics guard above

One fails against the pre-change code (the one asserting the argument is passed);
the other four hold both ways by design, since they pin behaviour that must be
preserved. Full CPU suite: 672 passed, 8 skipped.

### CODE REVIEW

**Comment by dvdimitrov13** [comment]
One follow-up I'd like your read on before doing anything about it.

`logits_to_keep=1` stops the head running over the whole sequence, but it does not stop `lm_head` being called, so the `2048 x 151936` weight matrix is still streamed on every forward. That is 622 MB in bf16, read to produce one position of logits nobody reads.

On a 5090 this does not matter. I profiled the weight traffic of the inference path and it comes to roughly 7 GB per forward against a ~3.9 ms bandwidth floor, while the measured time is ~63 ms, so the model is about 16x above the roofline and removing bytes buys nothing. On the bandwidth-bound deployment targets the picture changes: the same 622 MB is ~2.3 ms on Thor/Spark and ~3.0 ms on Orin AGX, and it is about 9% of all weight traffic in the model.

Replacing the head with an identity after load removes the read entirely. I checked and hidden states stay bitwise identical, since nothing downstream touches logits.

The catch is that it drops `lm_head.weight` from `state_dict`, so I do not think it belongs in `__init__` unconditionally. Two options I can see:

1. A `disable_lm_head()` method on `Qwen3Backbone` called from `Gr00tPolicy`, so it only applies on the inference path and training/checkpointing is untouched.
2. A config flag, defaulting off.

I lean towards 1, but I do not have a good sense of whether anything else in your deployment flow saves the backbone state dict after loading a policy. Would either be welcome as a follow-up PR, and is there a preference between them?

Happy to leave it alone if the state dict change is more trouble than the traffic is worth.
